gensim/models/doc2vec.py

```diff
@@ -794,6 +794,24 @@ def train(self, documents=None, corpus_file=None, total_examples=None, total_wor
 
         """
         kwargs = {}
+
+        # Check if both documents and corpus_file are None
+        if corpus_file is None and documents is None:
+            raise TypeError("Either one of corpus_file or documents value must be provided.")
+
+        # Check if both documents and corpus_file are not None
+        if corpus_file is not None and documents is not None:
+            raise TypeError("Instead provide value to either of corpus_file or documents parameter but not both.")
+
+        # Check if documents is not None and iterable but not string type
+        if documents is not None and isinstance(documents, Iterable) and isinstance(documents, string_types):
+            raise TypeError("Documents must be an iterable of list and not a string type.")
```

Comment:
I believe the test of is-`Iterable` and is-`string_types` is likely redundant, as all strings are iterable.

Reply:
I think it's also important to not go overboard with the checking. For example, you're making sure it's not a string here. Why not take that further and make sure it's not an int? A float? A file buffer? And so on.

Reply:
@gojomo I was not sure how to go about with handling this. `documents` should be an iterable 
but not a string since like you said strings are iterable. I think I got it now, so instead of checking for both, checking only for `string` type would do.

Reply:
Supplying a (path) string as `documents` might sufficiently reveal itself to be an error in other ways - like training that completes almost instantly, with little to no vocabulary. Separately, I'm not sure that all objects that would work because they are effectively "iterable" will actually test positive as being `isinstance()` `Iterable`.